microsoft/BotFramework-Composer
# Docker support + make bot runtime stateless

### What's include in this PR
* two Dockerfiles for composer and bot project, with pretty fine-tuned layers to make sure
   * the final image size is small (100~M), using multiple-stage image build
   * package installation can be cached to boost up docker image build speed
* a docker-compose.xml for docker-compose, with configuration for ports and share volume
* making the bot runtime stateless by accepting an zip file of bot
* refine the interface of bot connector to be just connect/reload instead of start\stop because we are assuming bot runtime should be started with composer
### How to use
Install docker-desktop for Windows/Linux/Mac

Run
* $ docker-compose up   // first time the image will be built, take about several minutes

Test Composer
* go to http://localhost:3000 to see Composer app

Test Bot
* start emulator and go to http://localhost:3979/api/messages

NOTE: we have to use ngrok in emulator to test bot, because it's logically a remote instance, and uncheck the "Bypass ngrok for local" in emulator settings. 

### Know issues
1. docker-image build will fail due to yarn install fails
solution: re-run docker compose up

### What's next
* Publish docker images
* The doc needs to updated
